log/loki.go

```diff
@@ -0,0 +1,422 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2020 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package log
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// TODO move this to it's own package
+// benchmark it
+
+type lokiHook struct {
+	addr             string
+	additionalParams [][2]string
+	ch               chan *logrus.Entry
+	limit            int
+	levels           []logrus.Level
+	pushPeriod       time.Duration
+	client           *http.Client
+}
+
+// LokiFromConfigLine returns a new logrus.Hook that pushes logrus.Entrys to loki and is configured
+// through the provided line
+func LokiFromConfigLine(line string) (logrus.Hook, error) {
+	h := &lokiHook{
+		addr:       "http://127.0.0.1:3100/loki/api/v1/push",
+		limit:      100,
+		levels:     logrus.AllLevels,
+		pushPeriod: time.Second * 1, // TODO configurable,
+		client: &http.Client{
+			Timeout: time.Second * 5, // TODO configurable, equal to pushPeriod?
+		},
+	}
+	if line == "loki" {
+		return h, nil
+	}
+
+	parts := strings.SplitN(line, "=", 2)
+	if parts[0] != "loki" {
+		return nil, fmt.Errorf("loki configuration should be in the form `loki=url-to-push` but is `%s`", line)
+	}
+	args := strings.Split(parts[1], ",")
+	h.addr = args[0]
+	// TODO use something better ... maybe
+	// https://godoc.org/github.com/kubernetes/helm/pkg/strvals
+	// atleast until https://github.com/loadimpact/k6/issues/926?
```

Comment:
Ugh, yeah... strvals or something like that sounds better than rolling out our own parsing logic everywhere... Even URL-encoded values might work: https://github.com/google/go-querystring

Reply:
Yeah, looking at the confusing mess of `,`, `;` and `=` signs that come from `additionalParams`, using a query string might not be a bad idea... https://github.com/gorilla/schema is another library that does it

Reply:
Though strvals is probably going to be the most readable of these... 
```
loki=somewhere:1233,labels.something=else,labels.foo=bar,limit=32,level=debug
```

Reply:
hmm btw it's probably going to be easy to write something like strvals, but which instead converts its format to JSON, which you then can easily unmarshal into a struct without bothering with reflection... but that can go in #883, for now strvals+mapstructure is probably the way to go... and I'll stop this conversation with myself now, so I can push the review :sweat_smile: